<center> <font size = 5> Waste water treatment plant data: Clustering analysis as a predictive mean </font> </center>

In [1]:
import pandas as pd
url = 'https://bd29ee0e-54ab-4daa-9671-d153865d1620.usrfiles.com/ugd/bd29ee_f0c5531356114a9ba2c93dc8965de700.csv'
df=pd.read_csv(url,header=None)
df.head()

0      1    2    3    4    5    6     7    8     9   ...    29    30  \
0  D-1/3/90  44101  1.5  7.8    ?  407  166  66.3  4.5  2110  ...  2000     ?   
1  D-2/3/90  39024    3  7.7    ?  443  214  69.2  6.5  2660  ...  2590     ?   
2  D-4/3/90  32229    5  7.6    ?  528  186  69.9  3.4  1666  ...  1888     ?   
3  D-5/3/90  35023  3.5  7.9  205  588  192  65.6  4.5  2430  ...  1840  33.1   
4  D-6/3/90  36924  1.5  8.0  242  496  176  64.8    4  2110  ...  2120     ?   

     31    32    33    34    35    36    37    38  
0  58.8  95.5     ?    70     ?  79.4  87.3  99.6  
1  60.7  94.8     ?  80.8     ?  79.5  92.1   100  
2  58.2  95.6     ?  52.9     ?  75.8  88.7  98.5  
3  64.2  95.3  87.3  72.3  90.2  82.3  89.6   100  
4  62.7  95.6     ?    71  92.1  78.2  87.5  99.5  

[5 rows x 39 columns]

In [2]:
df.shape

(527, 39)

This dataset is downloaded from UCI's Machine Learning Depository. Although it says on the website that the dataset has no missing value, as shown, that is not true. The dataset does have missing values.

Here is the cleaned dataset with the "?" replaced by the mean of the corresponding column. Python sort of treats the "?" as a string and would not calculate the mean just based on the the present values. To replace the "?" with the mean of the corresponding column, we need to sperate the "?" first. That is a lot of work and will take a while. Hence, the "?"'s have all been replaced in Excel, which just automatically ignores any non-numeric value and produce the mean based on the present numeric values. In addition, with a click, Excel can flash fill hundres of even thousands of columns in a second. There are some tasks that are much more convenient to perform in Excel.

In [3]:
url='https://bd29ee0e-54ab-4daa-9671-d153865d1620.usrfiles.com/ugd/bd29ee_57b940b192b3419aafaac279ea9c256b.csv'
df_original=pd.read_csv(url,header=None)
df_original.head()

0        1    2    3           4      5      6     7    8     9   \
0  D-1/3/90  44101.0  1.5  7.8  188.714286  407.0  166.0  66.3  4.5  2110   
1  D-2/3/90  39024.0  3.0  7.7  188.714286  443.0  214.0  69.2  6.5  2660   
2  D-4/3/90  32229.0  5.0  7.6  188.714286  528.0  186.0  69.9  3.4  1666   
3  D-5/3/90  35023.0  3.5  7.9  205.000000  588.0  192.0  65.6  4.5  2430   
4  D-6/3/90  36924.0  1.5  8.0  242.000000  496.0  176.0  64.8  4.0  2110   

   ...      29         30    31    32         33    34         35    36    37  \
0  ...  2000.0  39.085806  58.8  95.5  83.448049  70.0  89.013646  79.4  87.3   
1  ...  2590.0  39.085806  60.7  94.8  83.448049  80.8  89.013646  79.5  92.1   
2  ...  1888.0  39.085806  58.2  95.6  83.448049  52.9  89.013646  75.8  88.7   
3  ...  1840.0  33.100000  64.2  95.3  87.300000  72.3  90.200000  82.3  89.6   
4  ...  2120.0  39.085806  62.7  95.6  83.448049  71.0  92.100000  78.2  87.5   

      38  
0   99.6  
1  100.0  
2   98.5  
3  100.0  
4   99.5  

[5 rows x 39 columns]

The task for this dataset is clustering. We are going to use PCA to reduce the dimension before clustering. 

Drop the first column. It is just the dates.

In [4]:
df=df_original.drop(columns=0)
df.head()

1    2    3           4      5      6     7    8     9    10  ...  \
0  44101.0  1.5  7.8  188.714286  407.0  166.0  66.3  4.5  2110  7.9  ...   
1  39024.0  3.0  7.7  188.714286  443.0  214.0  69.2  6.5  2660  7.7  ...   
2  32229.0  5.0  7.6  188.714286  528.0  186.0  69.9  3.4  1666  7.7  ...   
3  35023.0  3.5  7.9  205.000000  588.0  192.0  65.6  4.5  2430  7.8  ...   
4  36924.0  1.5  8.0  242.000000  496.0  176.0  64.8  4.0  2110  7.9  ...   

       29         30    31    32         33    34         35    36    37  \
0  2000.0  39.085806  58.8  95.5  83.448049  70.0  89.013646  79.4  87.3   
1  2590.0  39.085806  60.7  94.8  83.448049  80.8  89.013646  79.5  92.1   
2  1888.0  39.085806  58.2  95.6  83.448049  52.9  89.013646  75.8  88.7   
3  1840.0  33.100000  64.2  95.3  87.300000  72.3  90.200000  82.3  89.6   
4  2120.0  39.085806  62.7  95.6  83.448049  71.0  92.100000  78.2  87.5   

      38  
0   99.6  
1  100.0  
2   98.5  
3  100.0  
4   99.5  

[5 rows x 38 columns]

In [5]:
from sklearn.decomposition import PCA
n_components = 10
pca = PCA(n_components=n_components,copy=True,whiten=True,svd_solver="randomized",random_state=0)
df_pca = pd.DataFrame(pca.fit_transform(df))
df_pca.head()

0         1         2         3         4         5         6  \
0 -1.052257  1.560858 -0.451688 -0.687762 -0.249067  0.345934  0.104893   
1 -0.272024  2.975218 -0.444741  0.203619 -0.021980 -0.858988 -1.027346   
2  0.767013  0.533639 -0.277885  1.438369  0.466340  0.866391 -1.291995   
3  0.340079  1.704061 -0.109157 -3.247605  0.207356  0.280321 -1.143034   
4  0.048455  1.642949 -0.350494  0.579759  0.650273 -1.217886 -1.213244   

          7         8         9  
0 -0.318714  0.035747 -0.676097  
1 -0.764313 -1.903119  0.671685  
2 -0.855199  0.177814 -0.824300  
3 -1.090338 -1.109255 -0.852072  
4 -0.923854 -0.476707 -0.054042

PCA has reduced the dimensionality of the dataset from 38 (excluding the date) to 10.

We are going to try the most well-known clustering algorithm KMeans first.

In [6]:
import numpy as np
from sklearn import metrics
from sklearn.cluster import KMeans
from sklearn.metrics import silhouette_score


n_clusters=np.arange(2,11)

for n_clusters in n_clusters:
    clusterer = KMeans(n_clusters=n_clusters, random_state=0)
    cluster_labels = clusterer.fit_predict(df_pca)
    silhouette_avg = silhouette_score(df_pca, cluster_labels,metric='euclidean')
    print("For n_clusters =", n_clusters,
          " ,the average silhouette_score is :", silhouette_avg,)

For n_clusters = 2  ,the average silhouette_score is : 0.16065526277517292
For n_clusters = 3  ,the average silhouette_score is : 0.0776368680227452
For n_clusters = 4  ,the average silhouette_score is : 0.0802051290200126
For n_clusters = 5  ,the average silhouette_score is : 0.09097937768668302
For n_clusters = 6  ,the average silhouette_score is : 0.09218695941711078
For n_clusters = 7  ,the average silhouette_score is : 0.08890189118452953
For n_clusters = 8  ,the average silhouette_score is : 0.10681533932314098
For n_clusters = 9  ,the average silhouette_score is : 0.10740066176647732
For n_clusters = 10  ,the average silhouette_score is : 0.10237577065022872


It seems that KMeans is not the right clustering algorithm for this dataset. Let's try Affinity Propagation andd evaluate the result with Davies-Bouldin Index.

In [7]:
from sklearn.cluster import AffinityPropagation
from sklearn.metrics import davies_bouldin_score


clusterer = AffinityPropagation(max_iter=600)
cluster_labels = clusterer.fit_predict(df_pca)
davies_bouldin_score = davies_bouldin_score(df_pca, cluster_labels)
score=silhouette_score(df_pca, cluster_labels,metric='euclidean')
print("The Davies Bouldin score is: ",davies_bouldin_score)
print("The Silhouette score is: ",score)

The Davies Bouldin score is:  1.176780836954748
The Silhouette score is:  0.07899551146471562


/opt/conda/envs/Python36/lib/python3.6/site-packages/sklearn/metrics/cluster/unsupervised.py:342: RuntimeWarning: divide by zero encountered in true_divide
  score = (intra_dists[:, None] + intra_dists) / centroid_distances
/opt/conda/envs/Python36/lib/python3.6/site-packages/sklearn/metrics/cluster/unsupervised.py:342: RuntimeWarning: invalid value encountered in true_divide
  score = (intra_dists[:, None] + intra_dists) / centroid_distances


Davies Bouldin Score is above 1.18, while we would want it to be close to 0. 

Mean-shift.

In [8]:
from sklearn.cluster import MeanShift, estimate_bandwidth

bandwidth = estimate_bandwidth(df_pca,quantile=0.95, n_jobs=-1)
ms = MeanShift(bandwidth=bandwidth, bin_seeding=True)
labels=ms.fit_predict(df_pca)
labels_unique = np.unique(labels)
n_clusters_ = len(labels_unique)
print("The number of clusters is: ",n_clusters_)
print("The silhouette score is: ",silhouette_score(df_pca, labels))

The number of clusters is:  4
The silhouette score is:  0.58606767392263


So far, MeanShift with band width gives us the best result.

Spectral clustering.

In [9]:
from sklearn.cluster import SpectralClustering

n_clusters=np.arange(2,13)

for n_clusters in n_clusters:
    clusterer = SpectralClustering(n_clusters=n_clusters, assign_labels="discretize",random_state=0)
    cluster_labels = clusterer.fit_predict(df_pca)
    silhouette_avg = silhouette_score(df_pca, cluster_labels,metric='euclidean')
    print("For n_clusters =", n_clusters,
          " ,the average silhouette_score is :", silhouette_avg,)

For n_clusters = 2  ,the average silhouette_score is : 0.6187113633890466
For n_clusters = 3  ,the average silhouette_score is : 0.5883341039007282
For n_clusters = 4  ,the average silhouette_score is : 0.03998756954394866
For n_clusters = 5  ,the average silhouette_score is : 0.030224482417288662
For n_clusters = 6  ,the average silhouette_score is : 0.031355738118854184
For n_clusters = 7  ,the average silhouette_score is : 0.054089634625681456
For n_clusters = 8  ,the average silhouette_score is : 0.023425913531352725
For n_clusters = 9  ,the average silhouette_score is : 0.04155135413263379
For n_clusters = 10  ,the average silhouette_score is : 0.036244323864623604
For n_clusters = 11  ,the average silhouette_score is : 0.03684314586503028
For n_clusters = 12  ,the average silhouette_score is : 0.051239919401743474


There is some hope here. SpectralClustering with 2 clusters seems to be the best algorithm at this moment.

In [10]:
from sklearn.cluster import AgglomerativeClustering

n_clusters=np.arange(2,13)
for n_clusters in n_clusters:
    clusterer = AgglomerativeClustering(n_clusters=n_clusters)
    cluster_labels = clusterer.fit_predict(df_pca)
    silhouette_avg = silhouette_score(df_pca, cluster_labels)
    print("For n_clusters =", n_clusters,
          " ,the average silhouette_score is :", silhouette_avg,)

For n_clusters = 2  ,the average silhouette_score is : 0.5958240072251978
For n_clusters = 3  ,the average silhouette_score is : 0.5625537747911646
For n_clusters = 4  ,the average silhouette_score is : 0.09543735974056813
For n_clusters = 5  ,the average silhouette_score is : 0.07047622316037176
For n_clusters = 6  ,the average silhouette_score is : 0.0808876136113748
For n_clusters = 7  ,the average silhouette_score is : 0.08477261796125093
For n_clusters = 8  ,the average silhouette_score is : 0.07769354356485994
For n_clusters = 9  ,the average silhouette_score is : 0.08237541229034659
For n_clusters = 10  ,the average silhouette_score is : 0.07804165879715264
For n_clusters = 11  ,the average silhouette_score is : 0.0812342699608606
For n_clusters = 12  ,the average silhouette_score is : 0.08757322253834249


Not as good as SpectralClustering.

In [11]:
from sklearn.cluster import DBSCAN

db = DBSCAN(eps=7).fit(df_pca)
labels = db.labels_
print("The number of clusters is: ",len(np.unique(labels)))
print("Silhouette score: %0.3f" % metrics.silhouette_score(df_pca, labels))

The number of clusters is:  2
Silhouette score: 0.708


Best result so far.

I am using IBM Watson Studio. For some unknown reason, I cannot import OPTICS. So, I will have to skip the OPTICS algorithm.

In [12]:
from sklearn.cluster import Birch

clusterer = Birch(threshold=0.9,branching_factor=100,n_clusters=2)
cluster_labels = clusterer.fit_predict(df_pca)
silhouette_avg = silhouette_score(df_pca, cluster_labels,metric='euclidean')
print("For n=2, the average silhouette_score is :", silhouette_avg,)

clusterer = Birch(threshold=0.9,branching_factor=100,n_clusters=3)
cluster_labels = clusterer.fit_predict(df_pca)
silhouette_avg = silhouette_score(df_pca, cluster_labels,metric='euclidean')
print("For n=3, the average silhouette_score is :", silhouette_avg,)

clusterer = Birch(threshold=0.9,branching_factor=100,n_clusters=4)
cluster_labels = clusterer.fit_predict(df_pca)
silhouette_avg = silhouette_score(df_pca, cluster_labels,metric='euclidean')
print("For n=4, the average silhouette_score is :", silhouette_avg,)

clusterer = Birch(threshold=0.9,branching_factor=100,n_clusters=5)
cluster_labels = clusterer.fit_predict(df_pca)
silhouette_avg = silhouette_score(df_pca, cluster_labels,metric='euclidean')
print("For n=5, the average silhouette_score is :", silhouette_avg,)

For n=2, the average silhouette_score is : 0.5958240072251978
For n=3, the average silhouette_score is : 0.5625537747911646
For n=4, the average silhouette_score is : 0.5609084576393202
For n=5, the average silhouette_score is : 0.10725783924705873


We have tried all available SKlearn clustering algorithms. **For this particular dataset and with the dimensionality of the dataset reduced to 10 by PCA, DBSCAN with eps=0.7 gives the best result.**

In [13]:
from sklearn.cluster import DBSCAN
from sklearn.metrics import davies_bouldin_score

db = DBSCAN(eps=7).fit(df_pca)
labels = db.labels_
print("The number of clusters is: ",len(np.unique(labels)))
print("Silhouette score is: %0.3f" % metrics.silhouette_score(df_pca, labels))
print("The Davies Bouldin score is: ",davies_bouldin_score(df_pca, labels))

The number of clusters is:  2
Silhouette score is: 0.708
The Davies Bouldin score is:  0.20454255118124065


/opt/conda/envs/Python36/lib/python3.6/site-packages/sklearn/metrics/cluster/unsupervised.py:342: RuntimeWarning: divide by zero encountered in true_divide
  score = (intra_dists[:, None] + intra_dists) / centroid_distances


Let's print out the two centroids.

In [14]:
from sklearn.neighbors import NearestCentroid

clf = NearestCentroid()
clf.fit(df_pca, labels)
clf.centroids_

array([[ 1.11953294e-01, -4.54943679e-01,  1.14545151e+01,
         2.65898293e-01, -2.32340268e+00,  7.37934571e-01,
        -7.96913948e-01,  1.72246220e+00,  1.53681804e+00,
         6.40549465e+00],
       [-2.12838961e-04,  8.64911937e-04, -2.17766448e-02,
        -5.05510062e-04,  4.41711536e-03, -1.40291744e-03,
         1.51504553e-03, -3.27464297e-03, -2.92170731e-03,
        -1.21777465e-02]])

In [15]:
labels=pd.DataFrame(labels,columns=["label"])
df_labeled=pd.concat([df_pca,labels],axis=1)
df_labeled.head()

0         1         2         3         4         5         6  \
0 -1.052257  1.560858 -0.451688 -0.687762 -0.249067  0.345934  0.104893   
1 -0.272024  2.975218 -0.444741  0.203619 -0.021980 -0.858988 -1.027346   
2  0.767013  0.533639 -0.277885  1.438369  0.466340  0.866391 -1.291995   
3  0.340079  1.704061 -0.109157 -3.247605  0.207356  0.280321 -1.143034   
4  0.048455  1.642949 -0.350494  0.579759  0.650273 -1.217886 -1.213244   

          7         8         9  label  
0 -0.318714  0.035747 -0.676097      0  
1 -0.764313 -1.903119  0.671685      0  
2 -0.855199  0.177814 -0.824300      0  
3 -1.090338 -1.109255 -0.852072      0  
4 -0.923854 -0.476707 -0.054042      0

In [16]:
np.unique(labels)

array([-1,  0])

In [17]:
print(len(df_labeled[df_labeled["label"]==0]),"instances belong to cluster 0.")
print(len(df_labeled[df_labeled["label"]==-1]),"instances belong to cluster 1.")

526 instances belong to cluster 0.
1 instances belong to cluster 1.


**It seems that after PCA, there is one instance that stands out from all others and clustering found it for us. This data was collected from a water treatment plant. This instance might represent a situation warrants particular attention. Let's print out this outlier.**

In [19]:
df_outlier=df_labeled[df_labeled["label"]==-1]
df_original.loc[df_outlier.index,:]

0        1     2    3      4      5       6     7         8   \
395  D-24/5/91  36495.0  0.12  7.7  213.0  627.0  2008.0  17.9  4.593825   

       9   ...      29    30    31       32    33    34    35    36    37  \
395  1257  ...  1289.0  68.5  96.1  90.5542  83.5  47.3  92.5  81.0  99.4   

           38  
395  99.08629  

[1 rows x 39 columns]

The instance that warrants the water treatment plant's attention occured on May 24, 1991. The above are the original recordings of the instance.

**That was an interesting application of clustering.** Usually, clustering analysis is used to find patterns. In this case, it found the pattern for us and also the instance that warrants the water treatment plant's attention. For this reason, this clustering algorithm could be used to predict faults in operational state.